File: dev/iRemote.Java/src/com/scientificsoft/iremote/server/hosts/HostsManager.java
/**
 * Copyright (c) 2013 Egor Pushkin. All rights reserved.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.scientificsoft.iremote.server.hosts;

import com.scientificsoft.iremote.platform.system.DateTimeUtils;
import com.scientificsoft.iremote.platform.tools.Vector;
import com.scientificsoft.iremote.server.config.Config;
import com.scientificsoft.iremote.server.protocol.Guid;

/**
 * HostsManager is not thread safe. The reason is that platform code 
 * does not require accessing it from multiple threads simultaneously.
 * It is utilized by client code immediately before establishing connection
 * and by platform immediately after connection is established.
 * Client code should not refer to HostsManager between DSession.connecting
 * and DSession.approved events.
 */
public class HostsManager {
    
    /** Holds single reference to class instance. */
    private static HostsManager instance_ = null;    
    
    /** List of known hosts. */
    private Vector hosts_ = new Vector();
    
    /** Last accessed host. */
    private NetInterface lastAccessed_ = new NetInterface("192.168.1.1", "8444");
    
    public static synchronized HostsManager instance() {
        if ( null == instance_ ) {
            instance_ = new HostsManager();
        }
        return instance_;
    }
    
    private HostsManager() {
    }    

    public NetInterface lastAccessed() {
    	return lastAccessed_;
    }
    
    public void setLastAccessed(String ip, String port) {
    	lastAccessed_.ip_ = ip;
    	lastAccessed_.port_ = port;
    	Config.save();
    }
    
    public int hostsCount() {
    	return hosts_.size();
    }
    
    public Host host(int index) {
    	if ( index < 0 || index >= hosts_.size() )
    		return null;
    	return (Host)hosts_.elementAt(index);    	
    }
    
    public Host host(Guid hostId) {
    	int amount = hosts_.size();
    	for ( int i = 0 ; i < amount ; ++i ) {
    		Host host = (Host)hosts_.elementAt(i);
    		if ( host.hostId().equals(hostId) )
    			return host;
    	}
    	return null;
    }
    
    public void addHost(Host host) {
    	try {
    		hosts_.add(host);
    	} catch ( Exception e ) {
    	}
    }
    
    public void removeInterface(int hostIndex, int interfaceIndex) {
    	try {
	    	// Check whether specified index exists in the list.
			if ( hosts_.size() <= hostIndex ) {
				return;
			}
			// Acquire host, which must be removed.
			Host host = (Host)hosts_.elementAt(hostIndex);		
			if ( host.interfacesCount() <= 1 ) {
				// Remove host with specified index.
				hosts_.removeElementAt(hostIndex);
			} else {
				// Remove host interface with specified index.
				host.removeInterface(interfaceIndex);
			}
	    	Config.save();
    	} catch ( Exception e ) {
    	}		
    }
    
	public void saveHost(Host srcHost) {
		// This locks users database internally for writing.
		try {
			// Check whether host record is ready to be saved. It should meet the 
			// Following conditions:
			// - host id must be specified;
			// - at least one interface must be added.
			if ( !srcHost.hasHostId() || 0 == srcHost.interfacesCount() ) {
				return;
			}		
            // Update last access time.
			srcHost.setLastAccessTime(DateTimeUtils.formatCurrentDate());
			// Remember first network interface of the host to be saved.
			NetInterface srcInterface = srcHost.interfaceAtIndex(0);
			// Find host by id.
			Host dstHost = host(srcHost.hostId());
			if ( null == dstHost ) {
	            // Add new host.
				addHost(srcHost);
	            Config.save();
				return;				
			}				
			// Find network interface in found host with the same ip and port.
			boolean interfaceFound = false;
			int dstInterfacesCount = dstHost.interfacesCount();
			for ( int i = 0 ; i < dstInterfacesCount ; ++i ) {
				NetInterface dstNetInterface = dstHost.interfaceAtIndex(i);
				if ( ( dstNetInterface.ip_.equalsIgnoreCase(srcInterface.ip_) ) && 
					( dstNetInterface.port_.equalsIgnoreCase(srcInterface.port_) ) ) {
					interfaceFound = true;
					break;
				}				
			}
			// Check whether interface with current ip and port is found.
			if ( !interfaceFound ) {
				// Add one more interface to found host.
				NetInterface netInterface = new NetInterface(srcInterface.ip_, srcInterface.port_);
				dstHost.addInterface(netInterface);
			}
				
			// Update records of found host.
            // Host object, which is added upon successful authentication (srcHost),
            // contains password, if
            // - it was entered by user (Holder::SubmitPassword);
            // - if was extracted from the db (ServerContext::HandlePasswordRequest);
            // and does not contain password if
            // - device was authorized anonymously (user allowed connection once).
            // Thus, password should be saved, if it exists, and should be 
            // cleared, otherwise.
			String password = srcHost.password();
            if ( null != password )
				dstHost.setMd5Password(password);			
            else
                dstHost.resetPassword();            
            // Update last access time.
            dstHost.setLastAccessTime(srcHost.lastAccessTime());
            Config.save();
		} catch ( Exception e ) {
		}		
	}
}

File: dev/iRemote.Cpp/Connector/Hosts/HostsManager.cpp
/**
 * Copyright (c) 2013 Egor Pushkin. All rights reserved.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "Common/Common.h"

#include "NetInterface.h"
#include "Host.h"
#include "HostsManager.h"

#include "../Config/Config.h"

namespace RemotePC
{

	HostsManager::HostsManager()
		: lock_( mc::Library::ReadWriteLock() )
		, hosts_()
	{
	}

	HostsManager& HostsManager::Instance()
	{
		return Loki::SingletonHolder< HostsManager >::Instance();
	}

	mc::IReadWriteLockPtr HostsManager::GetLock()
	{
		return lock_;
	}

	size_t HostsManager::GetHostsCount() const
	{
		return hosts_.size();
	}

	Host& HostsManager::GetHost(size_t index)
	{
		return hosts_.at(index);
	}
	
	Host& HostsManager::GetHost(mc::RefGuid hostId)
	{
		Hosts_::iterator iter = hosts_.begin();
		Hosts_::iterator end = hosts_.end();
		for ( ; iter != end ; ++iter )
		{		
			if ( (*iter).GetHostId() == hostId )
				return (*iter);
		}
		throw std::exception();		
	}
	
	bool HostsManager::HasHost(mc::RefGuid hostId)
	{
		Hosts_::iterator iter = hosts_.begin();
		Hosts_::iterator end = hosts_.end();
		for ( ; iter != end ; ++iter )
		{		
			if ( (*iter).GetHostId() == hostId )
				return true;
		}
		return false;
	}		

	void HostsManager::AddHost(const Host& host)
	{
		// This operation locks users database internally to write data.
		mc::IReadWriteLock::Writer_ locker(lock_);
		
		// Do add user.
		hosts_.push_back(host);
        
        // Save config.
        Config::Instance().Save();
	}

	void HostsManager::RemoveInterface(size_t hostIndex, size_t interfaceIndex)
	{
		// This operation locks users database internally to write data.
		mc::IReadWriteLock::Writer_ locker(lock_);
		
		// Check whether specified index exists in the list.
		if ( hosts_.size() <= hostIndex )
			return;

		// Acquire host, which must be removed.
		Host& host = hosts_.at(hostIndex);		
		if ( host.GetInterfacesCount() <= 1 )
		{
			// Remove host with specified index.
			hosts_.erase( hosts_.begin() + hostIndex);
		}
		else
		{
			// Remove host interface with specified index.
			host.RemoveInterface(interfaceIndex);
		}
        
        // Save config.
        Config::Instance().Save();
	}
	
	void HostsManager::SaveHost(const Host& srcHost) 
	{
		// This locks users database internally for writing.
		mc::IReadWriteLock::Writer_ locker(lock_);
		
		// Check whether host record is ready to be saved. It should meet the 
		// Following conditions:
		// - host id must be specified;
		// - at least one interface must be added.
		if ( !srcHost.HasHostId() || 0 == srcHost.GetInterfacesCount() )
			return;
		
		// Remember first network interface of the host to be saved.
		const NetInterface& srcInterface = srcHost.GetInterface(0);
		
		try
		{
			// Find host by id.
			Host& dstHost = HostsManager::Instance().GetHost(srcHost.GetHostId());				
			
			// Find network interface in found host with the same ip and port.
			bool interfaceFound = false;
			size_t dstInterfacesCount = dstHost.GetInterfacesCount();
			for ( size_t i = 0 ; i < dstInterfacesCount ; ++i )
			{
				const NetInterface& dstNetInterface = dstHost.GetInterface(i);
				if ( ( dstNetInterface.GetIp() == srcInterface.GetIp() ) && 
					( dstNetInterface.GetPort() == srcInterface.GetPort() ) )
				{
					interfaceFound = true;
					break;
				}				
			}
			// Check whether interface with current ip and port is found.
			if ( !interfaceFound )
			{
				// Add one more interface to found host.
				NetInterface netInterface;
				netInterface.SetIp(srcInterface.GetIp());
				netInterface.SetPort(srcInterface.GetPort());	
				dstHost.AddInterface(netInterface);
			}
			
			// Update records of found host.
            // Host object, whih is added upon successful authentication (srcHost),
            // contains password, if
            // - it was entered by user (Holder::SubmitPassword);
            // - if was extracted from the db (ServerContext::HandlePasswordRequest);
            // and does not contain password if
            // - device was authorized anonymously (user allowed connection once).
            // Thus, password should be saved, if it exists, and should be 
            // cleared, otherwise.
            if ( srcHost.HasPassword() )
				dstHost.SetMd5Password(srcHost.GetPassword());			
            else
                dstHost.ResetPassword();
		}
		catch ( ... )
		{
			// Host was not found. Add new host here. 
			AddHost(srcHost);
		}
        
        // Save config.
        Config::Instance().Save();
	}
	
}
